crates/argmin/src/solver/itp/itp_method.rs

```diff
@@ -0,0 +1,354 @@
+    #[error("ItpRoot error: tol must be positive.")]
+    NegativeTol,
+    /// tol must be nonzero
+    #[error("ItpRoot error: tol must be nonzero.")]
+    ZeroTol,
+    /// max must be larger than min
+    #[error("ItpRoot error: max must be larger than min.")]
+    MinLargerThanMax,
+}
+
+/// # ITP method
+///
+/// A root-finding algorithm, short for "interpolate, truncate, and project",
+/// that achieves superlinear convergence while retaining the worst-case
+/// performance of the bisection method.
+///
+/// ## Requirements on the optimization problem
+///
+/// The optimization problem is required to implement [`CostFunction`].
+///
+/// ##  Reference
+///
+/// [ITP Method]: https://en.wikipedia.org/wiki/ITP_Method
+/// [An Enhancement of the Bisection Method Average Performance Preserving Minmax Optimality]: https://dl.acm.org/doi/10.1145/3423597
+#[derive(Clone)]
+#[cfg_attr(feature = "serde1", derive(Serialize, Deserialize))]
+pub struct ItpRoot<F> {
+    /// required relative accuracy
+    tol: F,
+    /// tuned hyper-parameter 1, controls truncation size
+    kappa1: F,
+    /// tuned hyper-parameter 2, controls truncation size
+    kappa2: F,
+    /// tuned hyper-parameter slack variable, controls projection interval size
+    n0: F,
+    /// left boundary of current interval
+    a: F,
+    /// right boundary of current interval
+    b: F,
+    /// function value at `a`
+    fa: F,
+    /// function value at `b`
+    fb: F,
+    /// iteration counter, used in the projection step
+    j: F,
+    /// n_(1/2), a preprocessing variable
+    n1o2: F,
+    /// a preprocessing variable
+    nmax: F,
+}
+
+impl<F: ArgminFloat> ItpRoot<F> {
+    /// Constructor
+    /// The values `min` and `max` must be bracketing the root of the function.
+    /// The parameter `tol` specifies the relative error to be targeted.
+    /// The values `kappa1` and `kappa2` are hyper-parameters tuning the truncation size.
+    /// The parameter `n0` is a hyper-parameter slack variable controlling the projection interval
+    /// size.
+    pub fn new(min: F, max: F, tol: F, kappa1: F, kappa2: F, n0: F) -> Self {
+        ItpRoot {
+            tol,
+            kappa1,
+            kappa2,
+            n0,
+            a: min,
+            b: max,
+            fa: F::nan(),
+            fb: F::nan(),
+            j: F::zero(),
+            n1o2: F::nan(),
+            nmax: F::nan(),
+        }
+    }
+
+    /// Constructor with default hyperparameters
+    /// The values `min` and `max` must be bracketing the root of the function.
+    /// The parameter `tol` specifies the relative error to be targeted.
+    /// kappa1 is defaulted to 0.2 / (max - min).
+    /// kappa2 is defaulted to 2.0.
+    /// n0 is defaulted to 1.0.
+    pub fn from_defaults(min: F, max: F, tol: F) -> Self {
+        Self::new(
+            min,
+            max,
+            tol,
+            // kappa1, suggested from paper
+            float!(0.2) / (max - min),
```

Comment:
Wondering, maybe we want to have this function return `Result` and verify `max - min` cannot be zero so we don't panic here. Not sure though, maybe not worth it right now.